### ISSUE TITLE
Add sanitization to dynamic aws_resource_info labels to prevent metric parsing from breaking

### ISSUE THREAD

**Issue by migueleliasweb** [prometheus/cloudwatch_exporter#313]
Apparently the dynamic labels based on aws tags added to `aws_resource_info` are not sanitized. I recently found an issue where AWS resources that had a colon ":" in their tag names would cause the prometheus to fail parsing the exporter's output.

Here an example of the output that breaks metrics parsing:
```
echo "aws_resource_info{job="aws_sqs",instance="",arn="arn:aws:sqs:ap-southeast-2:123123123:my-events-queue",queue_name="my-events-queue",tag_Name="my-events-queue",tag_my:type="infra",tag_myTeam="myteam",} 1.0" | promtool check metrics
```

The error:
```
error while linting: text format parsing error in line 1: expected '"' at start of label value, found 'a'
```

I assume some other (less likely to happen) scenarios might also happen where a resource with a tag set to `foo="", my_sneaky_label="bar"` would also cause the exporter to output 2 tags: `tag_foo="", my_sneaky_label="bar`.

In summary, would it make sense to add a quick validation step to ensure the sanity of the dynamic label's names and values? Something like a regex where we only allow alphanumeric characters & numbers (maybe also dashes)?

**Comment by migueleliasweb** [comment]
I just found that apparently this step exists: https://github.com/prometheus/cloudwatch_exporter/blob/master/src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java#L596 but somehow apparently it didn't catch my usecase. 🤔

**Comment by brian-brazil** [comment]
https://github.com/prometheus/cloudwatch_exporter/blob/ef0a8acfe1ee5b3f44ba024ddc2b75701d6aea73/src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java#L673 is the relevant line, are you using the latest release?

> I also noticed that it's apparently not sanitizing the tag value.

There's no need to.

**Comment by migueleliasweb** [comment]
> https://github.com/prometheus/cloudwatch_exporter/blob/ef0a8acfe1ee5b3f44ba024ddc2b75701d6aea73/src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java#L673
> 
> is the relevant line, are you using the latest release?
> > I also noticed that it's apparently not sanitizing the tag value.
> 
> There's no need to.

I'm running the release 0.7 which in interestingly seems to replace colon `:` with an underscore the right way https://github.com/prometheus/cloudwatch_exporter/blob/1dcaf36e47a7ea2467c3f3b0a04b807c432d3987/src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java#L509 but in the other hand, the release 0.9 seems like it doesn't for some reason https://github.com/prometheus/cloudwatch_exporter/blob/3808a32a8159b86598124c8fca691335e87ac18d/src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java#L513

I also fail to understand why the tag values wouldn't need to be sanitized as they have the same potential problem of breaking the format. 🤔

**Comment by brian-brazil** [comment]
I think you've your versions the wrong way around.

> I also fail to understand why the tag values wouldn't need to be sanitized as they have the same potential problem of breaking the format.

Label values can be any valid UTF-8. Escaping of quotes is taken care of by the client library.

**Comment by migueleliasweb** [comment]
I apologise, I guess this problem was already fixed in the 0.8 release. See: https://github.com/prometheus/cloudwatch_exporter/pull/247/files

Due to my very **poor regex judging abilities** I think I misjudged what this line is doing: https://github.com/prometheus/cloudwatch_exporter/blob/3808a32a8159b86598124c8fca691335e87ac18d/src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java#L513. 

It seems like it's doing the right thing and replacing what does not matches that regex. 🤦 

I will make sure to upgrade to the latest release asap. Thanks for the support.
